executor/analyze.go

```diff
@@ -515,3 +526,83 @@ func (e *AnalyzeFastExec) buildStats() (hists []*statistics.Histogram, cms []*st
 	// TODO: do fast analyze.
 	return nil, nil, nil
 }
+
+type analyzeIndexIncrementalExec struct {
+	AnalyzeIndexExec
+	index *statistics.Index
+}
+
+func analyzeIndexIncremental(idxExec *analyzeIndexIncrementalExec) statistics.AnalyzeResult {
+	idx := idxExec.index
+	highBound := idx.Histogram.GetUpper(idx.Len() - 1)
+	values, err := codec.Decode(highBound.GetBytes(), len(idxExec.idxInfo.Columns))
+	if err != nil {
+		return statistics.AnalyzeResult{Err: err}
+	}
+	ran := ranger.Range{LowVal: values, HighVal: []types.Datum{types.MaxValueDatum()}}
+	hist, cms, err := idxExec.buildStats([]*ranger.Range{&ran}, false)
+	if err != nil {
+		return statistics.AnalyzeResult{Err: err}
+	}
+	oldHist, oldCMS, err := idx.RemoveUpperBound(idxExec.ctx.GetSessionVars().StmtCtx, values)
+	if err != nil {
+		return statistics.AnalyzeResult{Err: err}
+	}
+	hist, err = statistics.MergeHistograms(idxExec.ctx.GetSessionVars().StmtCtx, oldHist, hist, int(idxExec.maxNumBuckets))
+	if err != nil {
+		return statistics.AnalyzeResult{Err: err}
+	}
+	if oldCMS != nil && cms != nil {
+		err = cms.MergeCMSketch(oldCMS)
+		if err != nil {
+			return statistics.AnalyzeResult{Err: err}
+		}
+	}
+	result := statistics.AnalyzeResult{
+		PhysicalTableID: idxExec.physicalTableID,
+		Hist:            []*statistics.Histogram{hist},
+		Cms:             []*statistics.CMSketch{cms},
```

Comment:
Should we also check `oldCMS != nil && cms != nil` here?

Reply:
No, it can be nil actually.

Reply:
I mean, if oldCMS is not nil, while cms is nil, we would not do the merge, then we would update the CMSketch to be nil?

Reply:
But `cms` cannot be nil here.